File: cdk-deployment/cdk_deployment/cdk_deployment_stack.py
from aws_cdk import (
    Stack,
    RemovalPolicy,
    aws_s3 as s3,
    aws_cloudfront as cloudfront,
    aws_cloudfront_origins as origins,
    aws_s3_deployment as s3deploy,
    aws_iam as iam
)
from constructs import Construct

class CdkDeploymentStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        site_bucket = s3.Bucket(self, "aws-task-2-auto-deploy",
            removal_policy=RemovalPolicy.DESTROY,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            auto_delete_objects=True)

        oai = cloudfront.OriginAccessIdentity(self, "OAI_new",
            comment="OAI for my distribution"
        )

        distribution = cloudfront.Distribution(self, "MyStaticSiteDistribution",
            default_behavior=cloudfront.BehaviorOptions(
                origin=origins.S3Origin(site_bucket, origin_access_identity=oai),
                viewer_protocol_policy=cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS
            ),
            default_root_object="index.html")

        site_bucket.add_to_resource_policy(iam.PolicyStatement(
            actions=["s3:GetObject"],
            resources=[site_bucket.arn_for_objects("*")],
            principals=[iam.ServicePrincipal("cloudfront.amazonaws.com")],
            conditions={
                "StringEquals": {
                    "AWS:SourceArn": "arn:aws:cloudfront::{self.account}:distribution/{distribution.distribution_od}"
                }
            }
        ))
        site_bucket.grant_read(oai)

        s3deploy.BucketDeployment(self, "DeployWithInvalidation",
            sources=[s3deploy.Source.asset("../dist")],
            destination_bucket=site_bucket,
            distribution=distribution,
            distribution_paths=["/*"]
        )

File: cdk-deployment/lib/cdk-deployment-stack.ts
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';
import * as cf from "aws-cdk-lib/aws-cloudfront";
import * as iam from "aws-cdk-lib/aws-iam";

import { bucketName } from '../constants';

export class CdkDeploymentStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
  
      const websiteBucket = new s3.Bucket(this, 'WebsiteBucket', {
        bucketName,
        websiteIndexDocument: 'index.html',
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        autoDeleteObjects: true,
        publicReadAccess: false,
        blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      });

      const originAccessIdentity = new cf.OriginAccessIdentity(
        this,
        "WebAppBucketOAI",
        {
          comment: websiteBucket.bucketName,
        }
      );

      websiteBucket.addToResourcePolicy(new iam.PolicyStatement({
        actions:["S3:GetObject"],
        resources:[websiteBucket.arnForObjects("*")],
        principals:[new iam.CanonicalUserPrincipal(originAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId)]
      }))

      const cloudfront = new cf.CloudFrontWebDistribution(
        this,
        "WebAppDistribution",
        {
          originConfigs: [
            {
              s3OriginSource: {
                s3BucketSource: websiteBucket,
                originAccessIdentity: originAccessIdentity
              },
              behaviors: [{ isDefaultBehavior: true }],
            },
          ],
        }
      );
  
      new s3deploy.BucketDeployment(this, 'DeployWebsite', {
        sources: [s3deploy.Source.asset("../dist")],
        destinationBucket: websiteBucket,
        distribution: cloudfront,
        distributionPaths: ["/*"],
      });
  
      new cdk.CfnOutput(this, "DomainURL", {
        value: cloudfront.distributionDomainName,
      });
  }
}